src/main.rs

```diff
@@ -182,34 +257,38 @@ fn mem_parser(arg: &str) -> Result<u64, String> {
   Ok(base * scale)
 }
 
-fn load_files(files: &[String]) -> Arc<Mutex<Host>> {
-  let files: Vec<_> = files
+fn load_book(files: &[String], compile_opts: &CompileArgs) -> Book {
+  let mut book = files
     .iter()
     .map(|name| {
-      fs::read_to_string(name).unwrap_or_else(|_| {
+      let contents = fs::read_to_string(name).unwrap_or_else(|_| {
         eprintln!("Input file {:?} not found", name);
         process::exit(1);
+      });
+      contents.parse::<Book>().unwrap_or_else(|e| {
+        eprintln!("Parsing error {e}");
+        process::exit(1);
       })
     })
-    .collect();
-  let host = Arc::new(Mutex::new(host::Host::default()));
-  host.lock().unwrap().insert_def(
-    "HVM.log",
-    host::DefRef::Owned(Box::new(stdlib::LogDef::new({
-      let host = Arc::downgrade(&host);
-      move |wire| {
-        println!("{}", host.upgrade().unwrap().lock().unwrap().readback_tree(&wire));
-      }
-    }))),
-  );
-  for file_contents in files {
-    host.lock().unwrap().insert_book(&ast::Book::from_str(&file_contents).unwrap());
+    .fold(Book::default(), |mut acc, i| {
+      acc.nets.extend(i.nets);
+      acc
+    });
+  let transform_opts = TransformArg::opts_from_cli(&compile_opts.compile_opts);
+  if transform_opts.pre_reduce {
+    book
+      .pre_reduce(
+        &|x| compile_opts.pre_reduce_skip.iter().any(|y| x == y),
+        compile_opts.pre_reduce_memory as usize,
```

Comment:
you say in `mem_parser` that we want it as `u64` so the failure happens later, but the values are all `as`-casted, which will just truncate them. I think it would be better to error early there, rather than having to check & error everywhere downstream.